qek/kernel/kernel.py

```diff
@@ -15,26 +16,106 @@ class QuantumEvolutionKernel:
     """
 
-    def __init__(self, mu: float):
+    def __init__(self, mu: float, size_max: int | None = None):
         """Initialize the QuantumEvolutionKernel.
 
         Args:
             mu (float): Scaling factor for the Jensen-Shannon divergence
+            size_max (int, optional): If specified, only consider the first `size_max`
+                qubits of bitstrings. Otherwise, consider all qubits. You may use this
+                to trade precision in favor of speed.
         """
-        self.params: dict[str, Any] = {"mu": mu}
+        self.params: dict[str, Any] = {
+            "mu": mu,
+            "size_max": size_max,
+        }
         self.X: Sequence[ProcessedData]
         self.kernel_matrix: np.ndarray
 
     def __call__(
-        self, graph_1: ProcessedData, graph_2: ProcessedData, size_max: int | None = None
-    ) -> float:
+        self,
+        X1: Sequence[ProcessedData],
+        X2: Sequence[ProcessedData] | None = None,
+    ) -> NDArray[np.floating]:
+        """Compute a kernel matrix from two sequences of processed data.
+
+        This method computes a M x N kernel matrix from the Jensen-Shannon divergences
+        between all pairs of graphs in the two datasets. The resulting matrix can be used
+        as a similarity metric for machine learning algorithms.
+
+        If `X1` and `X2` are two sequences representing the processed data for a
+        single graph each, the resulting matrix can be used as a measure of similarity
+        between both graphs.
+
+        Args:
+            X1: processed data to be used as rows.
+            X2 (optional): processed data to be used as columns. If unspecified, use X1
+                as both rows and columns.
+        Returns:
+            np.ndarray: A len(X1) x len(X2) matrix where entry[i, j] represents the
+            similarity between rows[i] and columns[j], scaled by a factor that depends
+            on mu.
+        Notes:
+            The JSD is computed using the jensenshannon function from
+            `scipy.spatial.distance`, and it is squared because scipy function
+            `jensenshannon` outputs the distance instead of the divergence.
+        """
+        # If size is not specified, set it to the length of the largest bitstring.
+        size_max = self.params["size_max"]
+        if size_max is None:
+            if X2 is None:
+                # No need to walk the same source twice.
+                sources = [X1]
+            else:
+                sources = [X1, X2]
+            for source in sources:
+                for data in source:
+                    length = len(data.sequence.qubit_info)
+                    if size_max is None or size_max <= length:
+                        size_max = length
+
+        # Note: At this stage, size_max could theoretically still be `None``, if both `X1` and `X2`
+        # are empty. In such cases, `dist_excitation` will never be called, so we're ok.
+
+        mu = float(self.params["mu"])
```

Comment:
```suggestion
        mu: float = self.params["mu"]
```

Reply:
I preferred the conversion because I suspect that some users will use e.g. `numpy.float64` or some other variant of `float`. What do you think?